### ISSUE TITLE
Offline Capability

### ISSUE THREAD

**Issue by leo-neon** [snowplow-incubator/snowplow-flutter-tracker#31]
Hello Snowplow Team,

does the SDK support a caching logic when being offline? As far as I can tell, I cannot set timestamps on my own, the SDK takes care of it. Let's say I am offline and I am collecting some ScreenViews. As soon as I'm online again, we send the collected events to the collection endpoint. The automatically added timestamps would be incorrect, right?

**Comment by leo-neon** [comment]
If so, could you please add the functionality to add a timestamp manually?

**Comment by paulboocock** [comment]
Hi @leo-neon 

The tracker adds two timestamps by default, the first is the device clock of when the event was created. This happens as soon as you track any event. This is the `dvce_created_tstamp` in the warehouse.
The second timestamp is added to the event _just_ before it is emitted to the Snowplow collector. This is `dvce_sent_tstamp` in the warehouse.

Using these two timestamps is usually enough to figure out if an event sat on the device for a prolonged period. In the Snowplow pipeline we then use the difference of these two timestamps, along with the `collector_tstamp`, to calculate the `derived_tstamp` which is a best guess at when the event actually occured. By using `collector_tstamp` here, we can use the trusted the server clock rather than the potentially unreliable device clock.

You can also set the True Timestamp of the event if you wish (which is a timestamp that you 100% know is when this event actually occured at, often used for backend/server-side tracking). Currently, I'm not sure this tracker supports setting the True Timestamp on an event (@matus-tomlein could you confirm?). You should be careful with this property however, as setting it will override other timestamps in the warehouse (such as `derived_tstamp`) which might impact your downstream data models if they use this timestamp.

**Comment by matus-tomlein** [comment]
I can confirm that the tracker does not currently provide the option to set the custom true timestamp property. We may add this feature, but it is not clear to me how this would help you, @leo-neon. As Paul said, the tracker automatically sets the `dvce_created_tstamp` at the time when the events are tracked and also adds another timestamp when the events are sent. How would you use the true timestamp that is different than the already added ones?

**Comment by leo-neon** [comment]
Our use case is the offline logic. Whenever I am offline, I am still collecting data about the visited screens - not with snowplow but I have my own TrackingData Object. Since I'm offline, these TrackingData objects are added to a queue waiting to be sent when network connection is back. I set the time of event occurrence in this object. As soon as I am online again - let's say 5 minutes later -, I want to use the track() method of the Snowplow SDK to track all the screen views of the last five minutes. At this moment, I am using the ScreenView constructor and then I'll track the ScreenView via Snowplow. The timestamps which are now added automatically would not reflect the actual time when the event occurred, right?
